### ISSUE TITLE
nltk 3.9.0: Lookup error on import "Resource wordnet not found"

### ISSUE THREAD

**Issue by danjac** [nltk/nltk#3308]
Python 3.12, nltk 3.9

Following instructions to download corpora, immediately ran into this issue on either running `import nltk`  or `python -m nltk.download`

```
LookupError: 
**********************************************************************
  Resource wordnet not found.
  Please use the NLTK Downloader to obtain the resource:

  >>> import nltk
  >>> nltk.download('wordnet')
  
  For more information see: https://www.nltk.org/data.html

  Attempted to load corpora/wordnet
```

This was working in nltk `3.9b1`.

**Comment by michbsd** [comment]
Yep. I am seeing this too - it fails on the import - so no chance to download 'wordnet'


```
[...]
          import nltk
   File "/usr/local/lib/python3.9/site-packages/nltk/__init__.py", line 153, in <module>
     from nltk.translate import *
   File "/usr/local/lib/python3.9/site-packages/nltk/translate/__init__.py", line 24, in <module>
     from nltk.translate.meteor_score import meteor_score as meteor
   File "/usr/local/lib/python3.9/site-packages/nltk/translate/meteor_score.py", line 14, in <module>
     from nltk.stem.api import StemmerI
   File "/usr/local/lib/python3.9/site-packages/nltk/stem/__init__.py", line 34, in <module>
     from nltk.stem.wordnet import WordNetLemmatizer
   File "/usr/local/lib/python3.9/site-packages/nltk/stem/wordnet.py", line 13, in <module>
     class WordNetLemmatizer:
   File "/usr/local/lib/python3.9/site-packages/nltk/stem/wordnet.py", line 48, in WordNetLemmatizer
     morphy = wn.morphy
   File "/usr/local/lib/python3.9/site-packages/nltk/corpus/util.py", line 120, in __getattr__
     self.__load()
   File "/usr/local/lib/python3.9/site-packages/nltk/corpus/util.py", line 86, in __load
     raise e
   File "/usr/local/lib/python3.9/site-packages/nltk/corpus/util.py", line 81, in __load
     root = nltk.data.find(f"{self.subdir}/{self.__name}")
   File "/usr/local/lib/python3.9/site-packages/nltk/data.py", line 579, in find
     raise LookupError(resource_not_found)
 LookupError:
 **********************************************************************
   Resource wordnet not found.
   Please use the NLTK Downloader to obtain the resource:

   >>> import nltk
   >>> nltk.download('wordnet')

   For more information see: https://www.nltk.org/data.html

   Attempted to load corpora/wordnet

   Searched in:
     - '/root/nltk_data'
     - '/usr/local/nltk_data'
     - '/usr/local/share/nltk_data'
     - '/usr/local/lib/nltk_data'
     - '/usr/share/nltk_data'
     - '/usr/local/share/nltk_data'
     - '/usr/lib/nltk_data'
     - '/usr/local/lib/nltk_data'
 **********************************************************************
```

**Comment by kevinwuhoo** [comment]
I see the same issue ~~and pinning nltk to 3.8.1 seems to fix the issue for now~~. It looks like 3.9 just got released a few hours ago.

Edit don't do this because other comments mention that 3.8.1 have a security issue do this instead: https://github.com/nltk/nltk/issues/3308#issuecomment-2295213972.

**Comment by danjac** [comment]
`3.8.1` has a security issue, and `3.8.2` was deleted from PyPI for reasons.

I'm wondering how this got through CI, as it's such an obvious error. Although the CI does import the NLTK data, this step appears to be cached - so if it's already downloaded, it will work fine:  https://github.com/nltk/nltk/blob/develop/.github/workflows/ci.yaml#L53

In my case, I had a "clean" environment without any NLTK data. Maybe it would be a good idea **not** to cache the NLTK download step?

**Comment by Andrew5057** [comment]
As a workaround, `3.9b1` does not have the remote code execution problem or this circular dependency.

```
python3 -m pip install nltk==3.9b1
```

**Comment by danjac** [comment]
Yes, I pinned nltk to `3.9b1` as a temporary workaround.  Whatever the root cause of the circular dependency however, it might be a good idea not to cache the corpora download step in the CI as that would have caught the issue - saving a few seconds hardly seems worth it.

**Comment by stevenbird** [comment]
I suspect this https requirement is the culprit:
https://github.com/nltk/nltk/commit/344164bf8117fcdf751236d9427274d67ef585c9

**Comment by cclauss** [comment]
Proof: `python3 -c "import nltk; nltk.download('wordnet')"` fails.

**Comment by ekaf** [comment]
It is not a https problem: @michbsd's  [error log](https://github.com/nltk/nltk/issues/3308#issuecomment-2295193349l) clearly shows that the culprit is a cascade of imports (though not circular), starting with nltk.translate.meteor_score importing a stemmer, which imports WordNetLemmatizer, which needs the WordNet package. But since nltk could not finish loading, the downloader is not defined.

**Comment by cclauss** [comment]
How did this test pass on v3.9?
https://github.com/nltk/nltk/blob/980f435765ecc7bf27b41ebbb7dde6c375294a1d/.github/workflows/ci.yaml#L53

**Comment by ekaf** [comment]
The quick fix is to comment out the following two lines (L48 and 50) in _nltk/stem/wordnet.py_. Try it and the problem vanishes:

```
#    morphy = wn.morphy
 
#    _morphy = wn._morphy
```

**Comment by cclauss** [comment]
https://github.com/nltk/nltk/blob/980f435765ecc7bf27b41ebbb7dde6c375294a1d/nltk/stem/wordnet.py#L48-L50

https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-a-permanent-link-to-a-code-snippet

**Comment by ekaf** [comment]
Thanks @cclauss, of course the two code lines in my previous comment should have been marked up as code instead of a quote. It is corrected now, and there is even a more elaborate PR #3309 available. Please forgive my many mistakes.

**Comment by cclauss** [comment]
Nice work.  You have made no mistakes.

**Comment by zall-e** [comment]
> As a workaround, `3.9b1` does not have the remote code execution problem or this circular dependency.
> 
> ```
> python3 -m pip install nltk==3.9b1
> ```

awesome, thank you :)

**Comment by stevenbird** [comment]
3.9.1 is released

**Comment by danjac** [comment]
@stevenbird  Thank you for fixing this!

**Comment by amanmibra** [comment]
Thank you!

**Comment by ghost** [comment]
I have this issue on ArchLinux, `sudo pacman -S nltk-data `solved this

**Comment by aspoorna24** [comment]
I have faced the lookup error even after installing
'pip install nltk=3.9b1'
so the simple solution is to download WordNet Resource
`
import nltk
nltk.download('wordnet')
`

**Comment by cclauss** [comment]
https://github.com/nltk/nltk/tags Try v3.9.1?

**Comment by LinAGKar** [comment]
I'm still getting this issue with 3.9.1 on OpenSUSE Tumbleweed.

#3309 is not included in [3.9.1](https://github.com/nltk/nltk/commits/3.9.1/), it was merged into develop [right after](https://github.com/nltk/nltk/commits/6a981a8f2410fc030bd2058ee2cd2ec89172c1a7/) that tag.

**Comment by ryanamannion** [comment]
#3309 was included in 3.9.1, [the link you included](https://github.com/nltk/nltk/commits/6a981a8f2410fc030bd2058ee2cd2ec89172c1a7/) confirms that. Commits higher up are more recent. You can also confirm this for yourself by downloading the 3.9.1 dist and comparing `nltk/stem/wordnet.py` to [the changes here in the PR](https://github.com/nltk/nltk/pull/3309/files)

**Comment by LinAGKar** [comment]
>https://github.com/nltk/nltk/pull/3309 was included in 3.9.1, [the link you included](https://github.com/nltk/nltk/commits/6a981a8f2410fc030bd2058ee2cd2ec89172c1a7/) confirms that. Commits higher up are more recent.

Those commit lists merges together the non-linear commit graph into a linear list. But you can compare it with the other link I included, https://github.com/nltk/nltk/commits/3.9.1/, and see that "[Fix bug in WordNetLemmatizer](https://github.com/nltk/nltk/commit/7d1cbc71441f607daea0894333912268e2911cab)" isn't there. Or alternatively, you can look in `git log --graph 6a981a8f2410fc030bd2058ee2cd2ec89172c1a7`.

>You can also confirm this for yourself by downloading the 3.9.1 dist and comparing nltk/stem/wordnet.py to [the changes here in the PR](https://github.com/nltk/nltk/pull/3309/files)

So I have, and those changes are not included in https://github.com/nltk/nltk/archive/refs/tags/3.9.1.tar.gz

**Comment by ryanamannion** [comment]
Interesting, there seems to be a difference in the tarball you linked and the one hosted on pypi at https://pypi.org/project/nltk/#nltk-3.9.1.tar.gz

The pypi dist files do not have the wordnet import, neither in the tarball nor the wheel. The tarball hosted here on GitHub does have that import still. Are you installing nltk from pypi via pip?

**Comment by LinAGKar** [comment]
No, I'm installing the [openSUSE package](https://build.opensuse.org/package/show/openSUSE%3AFactory/python-nltk), which if I understand correctly is built by cloning the git repo and checking out the tag.

The [3.9.1](https://github.com/nltk/nltk/releases/tag/3.9.1) tag should probably point at d0c0e51cc6f317a9daa69e40e2d104ded9ef5de0, and not aca78cb2add4084f76b9eac921d8a73927d7a086. Of course, if the tag is changed, everyone who built from it will need to rebuild.

The tarball from PyPI might have been uploaded manually based on 980f435765ecc7bf27b41ebbb7dde6c375294a1d. I also note it doesn't contain the changelog for 3.9.1.

**Comment by ryanamannion** [comment]
I suggest you open a new Issue for this, given that the underlying problem seems to be the tag/ release and not the issue resolved via #3309

**Comment by ekaf** [comment]
[This earlier comment](https://github.com/nltk/nltk/issues/3308#issuecomment-2295253659) is rightly wondering how CI could pass. 
The reason is that CI currently runs on a full _nltk_data_ installation from cache. To prevent bugs like this in the future, CI should first test loading nltk and downloading a package, without any _nltk_data_ installed initially.
